packages/ember-application/lib/system/application.js

```diff
@@ -368,7 +380,23 @@ const Application = Engine.extend({
   buildInstance(options = {}) {
     options.base = this;
     options.application = this;
-    return ApplicationInstance.create(options);
+    let instance = ApplicationInstance.create(options);
+    this._applicationInstances.push(instance);
```

Comment:
I would vaguely prefer to have this done by the `ApplicationInstance`'s `init` method. In my head it pairs nicely with the fact that `ApplicationInstance` is now the one responsible for deregistering.

I don't feel super strongly... What do you think?

Reply:
Seems like we should make a small refactor to:

* `Application#_unwatchInstance` - does what `unwatchInstance` does now.
* `Application#_watchInstance` - is called by `ApplicationInstance#init` at the beginning of that method. Does the watching bit here.

Reply:
Sounds good to me!